### PR TITLE
[Sealing] Fix for emergency sealing panic

### PR DESCRIPTION
**Pull request by durkmurder**
### Context

As described [here](https://axiomzen.slack.com/archives/C015G65HR2P/p1624576255001800) in some situations consensus nodes are panicking.

Stack trace shows that error is somewhere in `Collect()` since consensus nodes behave correctly with 0 verification nodes but start panicking when there are `> 0` verification nodes. I suspect that it was caused by error in implementation of `Collect` which tries to produce an indexed array but can't actually do that if there aren't missing signatures for chunks.

Example: 

`AggregatedSignature` collected approval for chunk index > 0, then

```
for chunkIndex, sig := range as.signatures {
    aggregatedSigs[chunkIndex] = sig
}
```
will result in segfault since `chunkIndex >= len(aggregatedSigs)`

Other parts [rely](https://github.com/onflow/flow-go/blob/683fdd1209a87226e95aab8ec740ac10a2447608/module/validation/seal_validator.go#L274) on correct indexing of signatures so we can't make a list which doesn't contain ALL signatures.

I think that returning empty array in case we don't have all signatures is a sufficient solution.

Either way this situation will be impossible when we remove emergency sealing since if we don't have exactly `numberOfChunks` signatures it's an error.

### CODE REVIEW

File: engine/consensus/approvals/aggregated_signatures.go
Comment (smnzhu) [inline engine/consensus/approvals/aggregated_signatures.go:53]: Would it be better to return some error here instead of nil?

Comment (smnzhu) [reply engine/consensus/approvals/aggregated_signatures.go:53]: Is a [seal](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L105) still valid with no aggregated signature?

**Comment by smnzhu** [comment]
Do we need to update [HasSignature](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/aggregated_signatures.go#L40) as well?

**Comment by smnzhu** [comment]
On the other hand, is the change to `PutSignature` and `HasSignature` necessary in the first place? I'm just trying to understand how we could end up in a situation where it is called with a chunk index that is out of range?

Since [`numberOfChunks` is equal to the length of `chunkCollectors`](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L37-L44), shouldn't [this](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L137-L139) already take care of the check for us?

(btw, we should probably use `c.numberOfChunks` instead of `len(c.chunkCollectors)` in that check for clarity)

File: engine/consensus/approvals/aggregated_signatures.go
Comment (durkmurder) [reply engine/consensus/approvals/aggregated_signatures.go:53]: I don't think that we should error here, being no-op should be sufficient since it's more a check to support the invariant of having only valid indexes. Actual checks for chunk index range are performed way earlier before getting to this part.

> Is a [seal](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L105) still valid with no aggregated signature?

In terms of signatures count seal is valid iff it has a signature for every chunk. 
But since we have emergency sealing nodes can [skip](https://github.com/onflow/flow-go/blob/683fdd1209a87226e95aab8ec740ac10a2447608/module/validation/seal_validator.go#L249) validating this condition.

**Comment by durkmurder** [comment]
> Do we need to update [HasSignature](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/aggregated_signatures.go#L40) as well?
> 
> Because it seems like we call that before `PutSignature` [here](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L141-L151)

I don't this we need to update `HasSignature`, I presume you suggest to return an error but I would just to a soft return as it is right now since it doesn't really change the logic and the actual range check is done earlier

**Comment by durkmurder** [comment]
> On the other hand, is the change to `PutSignature` and `HasSignature` necessary in the first place? I'm just trying to understand how we could end up in a situation where it is called with a chunk index that is out of range?
> 
> Since [`numberOfChunks` is equal to the length of `chunkCollectors`](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L37-L44), shouldn't [this](https://github.com/onflow/flow-go/blob/smnzhu-patch-1/engine/consensus/approvals/approval_collector.go#L137-L139) already take care of the check for us?
> 
> (btw, we should probably use `c.numberOfChunks` instead of `len(c.chunkCollectors)` in that check for clarity)

We need to update `PutSignature` to make sure that we don't include any signatures which are `chunkIndex >= numberOfChunks` since invariant and logic of `Collect()` [depend](https://github.com/onflow/flow-go/pull/896/files#diff-6afdcbf4dbcb85c0ea6d6f2fc987b1d5004965559f886d966a78f2e9ad688c4dR57) on that. Currently it's guaranteed by external logic but we also need to be sure that we do some internal checks. Something like [this](https://github.com/onflow/flow-go/pull/896/files#diff-76594a8775b7cc1dbc004155ee7c44586a3802950db2fd06959034b25eb5250dR19) easily adds invalid chunkIndex breaking invariant.

File: engine/consensus/approvals/aggregated_signatures.go
Comment (AlexHentschel) [reply engine/consensus/approvals/aggregated_signatures.go:53]: returning `nil` here: this would essentially introduce two different conventions: 
1. if we require approvals, we expect that there is an aggregated signature for each chunk 
2. if we don't require approvals, there are no aggregated signatures 

My preference would be to stick to a uniform convention that naturally covers both cases:
  * the seal should contain an aggregated signature for each chunk (in all cases).
  * for each chunk, the number of approvers must be larger or equal to `requiredApprovalsForSealVerification`. I.e. when we don't require any approvals, we include an empty `AggregatedSignatures`.

Our current implementation of the `sealValidator` accepts both conventions: https://github.com/onflow/flow-go/blob/77cd71b7930a2016fa09a1b79a2afa459b237903/module/validation/seal_validator.go#L249-L258 Nevertheless, my preference would be (on the generating side) to stick with one uniform convention that includes the degenerate edge case. Thereby, the `sealValidator` is flexible in what it accepts, but we generate the seals according to a single convention.

**Review by AlexHentschel — COMMENTED** [review]
Thanks Yurii, the diagnosis is great and the tests clearly show where the problem is: https://github.com/onflow/flow-go/blob/54367bdf3506a326eb8bf9591b0eb686d75b6d97/engine/consensus/approvals/aggregated_signatures.go#L46

I feel that the best approach would be to consolidate our convention how we aggregate approvals. My preference would be to stick to a uniform convention that naturally covers all cases:
  * the seal should contain an aggregated signature for each chunk (in all cases).
  * for each chunk, the number of approvers must be larger or equal to `requiredApprovalsForSealVerification`. I.e. when we don't require any approvals, we include an empty `AggregatedSignatures`.

In contrast, we currently have completely different conventions whether or not we require approvals. I have created a PR with my suggestions: https://github.com/onflow/flow-go/pull/899 It should be passing all tests 🤞

**Review by AlexHentschel — APPROVED** [review]
### I think Yurii’s PR _in its current form_ would be a great hot-patch for TestNet & Mainnet (_without_ my recommendations from [PR 899](https://github.com/onflow/flow-go/pull/899))


For the long-term, I’d like to slightly extend the scope of the patch, which requires also adjusting the [sealValidator](https://github.com/onflow/flow-go/blob/54367bdf3506a326eb8bf9591b0eb686d75b6d97/module/validation/seal_validator.go#L249-L258). Thanks Yurii for pointing this out 🙇 . Context: 
* For example, consider a block with 10 chunks, where only chunk 4 didn’t get sufficient approvals. When we emergency seal with the hot-patch logic (this PR _in its current form_), we don’t include any approvals.
* I’d prefer if we included the known approvals, even if we emergency seal. The change in logic is relatively straight forward (partially implemented in my  [PR 899](https://github.com/onflow/flow-go/pull/899) but needs additional revisions on sealValidator).

**Review by zhangchiqing — APPROVED** [review]

**Comment by durkmurder** [comment]
I am closing this one since it was already cherry picked to [908](https://github.com/onflow/flow-go/pull/908). 
[This](https://github.com/onflow/flow-go/pull/899) extended PR will replace current one.

**Pull request closed without merging**
